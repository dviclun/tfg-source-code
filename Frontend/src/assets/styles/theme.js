import { createTheme } from "@mui/material";
import { orange } from '@mui/material/colors';
import Poppins from '../fonts/Poppins-Regular.ttf'

export const theme = createTheme({
    status: {
        danger: orange[500],
      },
    palette: {
        slate100: '#f1f5f9',
        slate300:'#cbd5e1',
        green300: '#86efac',
        green400: '#4ade80',
        green500: '#22c55e',
        green600: '#16a34a',
        green700: '#047857', 
        green900: '#14532d',
        red400: '#f87171',
        red500: '#ef4444',
        zinc200: '#e4e4e7',
        zinc300: '#d4d4d8',
        zinc400: '#a1a1aa',
        rose400: '#fb7185',
        rose500: '#f43f5e',
        purple400: '#c084fc',
        purple500: '#a855f7',
        blue300: '#93c5fd',
        blue400: '#60a5fa',
        gray100: '#f3f4f6',
        neutral100: '#f5f5f5',
        neutral200: '#e5e5e5',
        emerald100: '#d1fae5'
    },
});