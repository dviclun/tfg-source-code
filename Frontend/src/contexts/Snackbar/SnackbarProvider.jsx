
import { useState } from 'react'
import { SnackbarContext } from './SnackbarContext';
import { Alert, Snackbar } from '@mui/material';
import { useTranslation } from 'react-i18next';

export const SnackbarProvider = ({children}) => {

    const [openSnackbar, setOpenSnackbar] = useState(false);
    const [openErrorSnackbar, setOpenErrorSnackbar] = useState(false);
    const [openWarningSnackbar, setOpenWarningSnackbar] = useState(false);
    const [message, setMessage] = useState('');
    const [warningMessage, setWarningMessage] = useState('');

    const [t] = useTranslation("global");

    const handleCloseSnackbar = (event, reason) => {
        if (reason === 'clickaway'){
            return;
        }

        setOpenSnackbar(false);
    }

    const handleCloseErrorSnackbar = (event, reason) => {
        if (reason === 'clickaway'){
            return;
        }

        setOpenErrorSnackbar(false);
    }

    const handleCloseWarningSnackbar = (event, reason) => {
        if (reason === 'clickaway'){
            return;
        }

        setOpenWarningSnackbar(false);
    }

    const handleOpenSnackbar = () => {
        setOpenSnackbar(true);
    }

    const handleOpenErrorSnackbar = () => {
        setOpenErrorSnackbar(true);
    }

    const handleOpenWarningSnackbar = () => {
        setOpenWarningSnackbar(true);
    }

  return (
    <SnackbarContext.Provider value={{
        setMessage,
        setWarningMessage,
        handleOpenSnackbar,
        handleOpenErrorSnackbar,
        handleOpenWarningSnackbar
    }}>
        {children}
        <Snackbar
        open={openSnackbar}
        autoHideDuration={4000}
        onClose={handleCloseSnackbar}
      >
        <Alert
            onClose={handleCloseSnackbar}
            severity="success"
            variant="filled"
            sx={{ width: '100%' }}
        >
        {message}
        </Alert>
      </Snackbar>

      <Snackbar
        open={openErrorSnackbar}
        autoHideDuration={3000}
        onClose={handleCloseErrorSnackbar}
     >
        <Alert
            onClose={handleCloseErrorSnackbar}
            severity='error'
            variant='filled'
            sx={{ width: '100%' }}
        >
            {t("serverSideErrorMessage")}
        </Alert>
     </Snackbar>  

     <Snackbar
        open={openWarningSnackbar}
        autoHideDuration={3000}
        onClose={handleCloseWarningSnackbar}
     >
        <Alert
            onClose={handleCloseWarningSnackbar}
            severity='warning'
            variant='filled'
            sx={{ width: '100%' }}
        >
            {warningMessage}
        </Alert>
     </Snackbar>  
    </SnackbarContext.Provider>
  )
}
