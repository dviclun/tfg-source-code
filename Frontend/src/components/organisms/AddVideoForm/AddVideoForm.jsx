import { Alert, Button, FormControl, Grid, Input, InputLabel, MenuItem, Select, Snackbar, Typography } from "@mui/material"

import AddToQueueIcon from '@mui/icons-material/AddToQueue';
import { useTheme } from '@mui/material/styles';
import { useContext, useEffect, useState } from "react";
import { useTranslation } from "react-i18next";
import { SnackbarContext } from "../../../contexts/Snackbar/SnackbarContext";
import { Header } from "../../molecules/Header/Header";



export const AddVideoForm = () => {

    const theme = useTheme();  
    
    //Estados por defecto
    const [videoNameEs, setVideoNameEs] = useState('');
    const [videoNameEn, setVideoNameEn] = useState('');
    const [videoUrl, setVideoUrl] = useState('');
    const [muscularGroup, setMuscularGroup] = useState('');
    const [difficulty, setDifficulty] = useState('');
    const [mcGroups, setMcGroups] = useState([]);

    const {handleOpenErrorSnackbar} = useContext(SnackbarContext);

    const [openSnackBar, setOpenSnackbar] = useState(false);

    const [t, i18n] = useTranslation("global");

    const [formErrors, setFormErrors] = useState(false);
    const [urlError, setUrlError] = useState(false);

    const FETCH_URL = 'https://tfg-backend-seven.vercel.app/muscularGroups';

    const urlRegex = /https?:\/\/(www\.)?[-a-zA-Z0-9@:%._\+~#=]{1,256}\.[a-zA-Z0-9()]{1,6}\b([-a-zA-Z0-9()@:%_\+.~#?&//=]*)/gi;

    useEffect(()=> {
        if(mcGroups.length === 0){
            fetch(FETCH_URL)
              .then(res => res.json())
              .then(data => {
                setMcGroups(data);
              })
              .catch(error => {
            console.log(error);
          })
          }
    }, [])

    const handleVideoNameEs = (e) => {
        setVideoNameEs(e.target.value);
    }

    const handleVideoNameEn = (e) => {
        setVideoNameEn(e.target.value);
    }

    const handleVideoUrl = (e) => {
        setVideoUrl(e.target.value);
    }

    const handleMuscularGroup = (e) => {
        setMuscularGroup(parseInt(e.target.value));
    }

    const handleDifficulty = (e) => {
        setDifficulty(e.target.value);
    }

    const handleFormSubmit = () => {
        if(videoNameEs.trim() !== '' && videoNameEn.trim() !== '' && videoUrl.trim() !== '' && muscularGroup !== '' && difficulty !== ''){
            setFormErrors(false);

            if(videoUrl.match(urlRegex)){
                setUrlError(false);
                addVideo();
            } else {
                setUrlError(true);
                return;
            }  
        } else {
            setFormErrors(true);
        }
    }

    const addVideo = () => {
        //Comprobacion de que los datos enviados al back son del tipo que queremos
        if(typeof videoNameEs === 'string' &&  typeof videoNameEn === 'string' && typeof videoUrl === 'string' && typeof muscularGroup === 'number' && typeof difficulty === 'string' ){
            
            let reqBody = {
                video_title_es: videoNameEs,
                video_title_en: videoNameEn,
                video_url: videoUrl,
                muscular_group: muscularGroup,
                difficulty: difficulty
            }
    
            let fetchConfig = {
                method: 'POST',
                headers: {
                    'Content-Type': 'application/json'
                },
                body: JSON.stringify(reqBody)
            }
    
            fetch('https://tfg-backend-seven.vercel.app/videos', fetchConfig)
                        .then(res => res.json())
                        .then(res => {
                            console.log('Video insertado');
                            clearForm();
                            handleOpenSnackbar();
                        })
                        .catch(error => {
                            handleOpenErrorSnackbar();
                        })
        } else {
            handleOpenErrorSnackbar();
        }
        
    }

    const handleOpenSnackbar = () => {
        setOpenSnackbar(true);
    }

    const handleCloseSnackbar = () => {
        setOpenSnackbar(false);
    }

    const clearForm = () => {
        setVideoNameEs('');
        setVideoNameEn('');
        setVideoUrl('');
        setMuscularGroup('');
        setDifficulty('');
    }

    const labelStyle = {
        fontFamily: 'Poppins',
        fontSize: '14px',
        "&.Mui-focused": {
            color: theme.palette.green400
        },        
    }

    const inputStyle = {
        marginTop: '30px',
        ':after': {
            borderBottomColor: theme.palette.green400
        }
    }

    const spanishSorted = [...mcGroups].sort((a,b) => a.group_name_es.localeCompare(b.group_name_es));
    const englishSorted = [...mcGroups].sort((a,b) => a.group_name_en.localeCompare(b.group_name_en));

  return (
    <>
        <Header />
        <Grid sx={{marginTop: '80px', padding: '2em'}}>
            <Grid item xs={12}>
                <Typography variant="h6" sx={{display: 'flex', alignItems: 'center', fontFamily:'Poppins'}}><AddToQueueIcon sx={{marginRight: '5px'}} />{t("videoGallery.addNewVideo")}</Typography>
                <hr/>
            </Grid>

            <Grid justifyContent={'center'} container item xs={12} sx={{marginTop: '1em'}} spacing={5}>
                <Grid item xs={12} md={4}>
                    <FormControl sx={{paddingBottom: '5%', width: '100%'}}>
                        <InputLabel sx={labelStyle}  htmlFor='videoNameEs'>{t("videoGallery.spanishVideoName")}</InputLabel>
                        <Input sx={inputStyle}  id="videoNameEs" autoComplete="off" name="videoNameEs" onChange={handleVideoNameEs} value={videoNameEs} />
                    </FormControl>
                </Grid>
                <Grid item xs={12} md={4}>
                    <FormControl sx={{paddingBottom: '5%', width: '100%'}}>
                        <InputLabel sx={labelStyle}  htmlFor='videoNameEn'>{t("videoGallery.englishVideoName")}</InputLabel>
                        <Input sx={inputStyle}  id="videoNameEn" autoComplete="off" name="videoNameEn" onChange={handleVideoNameEn} value={videoNameEn}/>
                    </FormControl>
                </Grid>
                <Grid item xs={12} md={4}>
                    <FormControl sx={{paddingBottom: '5%', width: '100%'}}>
                        <InputLabel sx={labelStyle} htmlFor='videoUrl'>{t("videoGallery.videoUrl")}</InputLabel>
                        <Input sx={inputStyle}  id="videoUrl" autoComplete="off" name="videoUrl" onChange={handleVideoUrl} value={videoUrl}/>
                    </FormControl>
                </Grid>
                
                <Grid item xs={12} md={4}>

                    <FormControl variant="standard" sx={{width: '100%'}}>
                        <InputLabel sx={labelStyle} id="muscularGroupLabel">{t("videoGallery.muscularGroup")}</InputLabel>
                        <Select
                            sx={inputStyle}
                            labelId="muscularGroupLabel"
                            id="muscularGroup"
                            label="Muscular Group"
                            onChange={handleMuscularGroup}
                            value={muscularGroup}
                     
                        >
                            {
                                (i18n.language === 'es')
                                ? spanishSorted.map(group =>(
                                    <MenuItem key={group.group_id} value={group.group_id}>{group.group_name_es}</MenuItem>
                                  ))
                                : englishSorted.map(group =>(
                                    <MenuItem key={group.group_id} value={group.group_id}>{group.group_name_en}</MenuItem>
                                  )) 
                            }

                        </Select>
                    </FormControl>
                
                </Grid>

                <Grid item xs={12} md={4}>

                    <FormControl variant="standard" sx={{width: '100%'}}>
                        <InputLabel sx={labelStyle} id="difficultyLabel">{t("videoGallery.difficulty")}</InputLabel>
                        <Select
                            sx={inputStyle}
                            labelId="difficultyLabel"
                            id="difficulty"
                            label="Difficulty"
                            onChange={handleDifficulty}
                            value={difficulty}
                       
                        >
                            <MenuItem value="starter">{t("videoGallery.starter")}</MenuItem>
                            <MenuItem value="medium">{t("videoGallery.medium")}</MenuItem>
                            <MenuItem value="advanced">{t("videoGallery.advanced")}</MenuItem>

                        </Select>
                    </FormControl>

                </Grid>
                {
                    (formErrors)
                    ? <Grid container justifyContent={'center'} item xs={12}>
                        <Typography><span className="errorSpan mt-0">{t("loginPage.allFieldsRequired")}</span></Typography>
                      </Grid>
                    : null
                }
                {
                    (urlError)
                    ? <Grid container justifyContent={'center'} item xs={12}>
                        <Typography><span className="errorSpan mt-0">{t("videoGallery.invalidUrl")}</span></Typography>
                      </Grid>
                    : null
                }

                <Grid container justifyContent={'center'} item xs={12}>
                    <Button onClick={handleFormSubmit} variant="contained" className="greenGradient text-white" sx={{fontFamily: 'Poppins', width: '25%', fontWeight: 'bold', minWidth: '150px'}}>{t("videoGallery.saveTraining")}</Button>
                </Grid>
                    
            </Grid>
        </Grid>

        <Snackbar open={openSnackBar} autoHideDuration={3000} onClose={handleCloseSnackbar}>
        <Alert
          onClose={handleCloseSnackbar}  
          severity="success"
          variant="filled"
          sx={{ width: '100%' }}
        >
          {t("videoGallery.feedback")}
        </Alert>
      </Snackbar>
    </>
  )
}
