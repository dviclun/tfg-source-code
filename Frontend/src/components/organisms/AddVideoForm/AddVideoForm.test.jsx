import { render } from "@testing-library/react";
import { BrowserRouter } from "react-router-dom";
import { AuthProvider } from "../../../contexts/Auth/AuthProvider";
import { SnackbarProvider } from "../../../contexts/Snackbar/SnackbarProvider";
import { ThemeProvider } from "@emotion/react";
import { theme } from "../../../assets/styles/theme";
import { I18nextProvider } from "react-i18next";
import i18next from "i18next";
import { AddVideoForm } from "./AddVideoForm";

describe('Test in <AddVideoForm/>', () => {
    test('Container must render and match snapshot', () => {

        const { container } = render(
            <BrowserRouter>
                <AuthProvider>
                    <SnackbarProvider>
                        <ThemeProvider theme={theme}>
                            <I18nextProvider i18n={i18next}>
                                <AddVideoForm />
                            </I18nextProvider>
                        </ThemeProvider>
                    </SnackbarProvider>
                </AuthProvider>
            </BrowserRouter>
        )

        expect(container).toBeTruthy();
        expect(container).toMatchSnapshot();
    })
})