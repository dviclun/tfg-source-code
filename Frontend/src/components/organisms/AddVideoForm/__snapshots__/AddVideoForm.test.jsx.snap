// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test in <AddVideoForm/> Container must render and match snapshot 1`] = `<div />`;
